### PR TITLE
Add algorithm to aggregate stacks images from wavelength dependend data

### LINKED ISSUE #13839: Algorithm to aggregate stacks for images for multiple energy levels
This is what at first the "energy bands" tab of the Tomo GUI will do (#13467). This should support the basic options of that tab.

Take a multi energy band stack (thousands of bands per projection), and aggregate into a simple (single energy level) stack (one image per projection).

This could be called `ImggAggregateEBands` or similar.

For a start, work with the indices of the energy bands/levels. Soonish we should get information on ToF/energy and potential gaps in energy levels from the FITS headers. What that's sorted out, the algorithm should be extended to use energy information appropriately.

### PR DESCRIPTION
**Pull request by FedeMPouzols**
Add the algorithm as described in #13839. Earlier, python scripts had been added in #14095.
Once the algorithm is available, it will be used from the tomo GUI (#13467).

**To test:**
- Ensure that the new unit and system tests pass. As this is a purely  I/O algorithm the unit tests don't do any real functional test (would be slow). The system test does.
- Check that the documentation reads well.

Open up the algorithm dialog and set:
- InputPath to the directory where you put the files that (at ISIS)  you can grab a small one from \\olympic\Babylon5\Public\fedemp\IMAT\test_stack_test_larmor_201510_one_angle_smallish
- OutputPath to your favorite new/empty directory.
- Run the algorithm and see that what it generates makes sense.
- It has to add quite a few files so it will take some seconds. The progress bar should indicate progress
- There are some bogus directories and files included in the test data, these should be ignored and you might see warnings about it

More points to bear in mind:
- This produces outputs as fits images. You can open them with imageJ, fv, or with the Mantid Tomo GUI.
- If you set the "Log Level" to information or more, you should see messages like "Saved output aggregated image into: /home/fedemp/test/test_imgg_agg/sum_projection_00001_bands_indices_0_1000"
  by the end of the algorithm run.
- For a more intensive (and slow) test, you can grab these other stacks of images as input: 
  - \\olympic\Babylon5\Public\fedemp\IMAT\test_stack_test_larmor_201510_one_angle
  - \\olympic\Babylon5\Public\fedemp\IMAT\test_stack_test_phantom1_3angles

Fixes #13839.

Release notes: https://github.com/mantidproject/mantid/commit/3d2b31144a1434ea1a842e0300f64b1ab33397ae

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/ImggAggregateWavelengths.cpp [hunk 1]
```diff
@@ -0,0 +1,834 @@
...
+                                         "normally it is much smaller. Please "
+                                         "check that you are providing the "
+                                         "correct input parameters");
+  }
+
+  for (size_t count = 0; count < availableCount; count += inc) {
+    size_t max =
+        ((count + inc) > availableCount) ? availableCount : count + inc;
+    // TODO call the 'ranges' version with the convenient range
+    ranges.emplace_back(std::make_pair(count, max));
```
Comment (quantumsteve) [inline Framework/DataHandling/src/ImggAggregateWavelengths.cpp]: With `emplace_back` you don't need to use `std::make_pair`. See, for example, #15874.

File: Framework/DataHandling/src/ImggAggregateWavelengths.cpp [hunk 2]
```diff
@@ -0,0 +1,834 @@
...
+    if (2 != minMaxTokens.count()) {
+      throw std::invalid_argument(
+          "Could not parse a minimum and maximum value separated by '" + sep +
+          "' from the string: " + str);
+    }
+
+    try {
+      ranges.emplace_back(
+          std::make_pair(boost::lexical_cast<size_t>(minMaxTokens[0]),
+                         boost::lexical_cast<size_t>(minMaxTokens[1])));
```
Comment (quantumsteve) [inline Framework/DataHandling/src/ImggAggregateWavelengths.cpp]: With `emplace_back` you don't need to use `std::make_pair`. See, for example, #15874.

**Comment by shahrozahmed1** [comment]
Ran `ImggAggregateWavelengths` system, documentation and unit test locally; which all passed. Compressive documentation which covers the weakness and strength of the algorithm and other useful information. 

Tested `.fits` files in `test_stack_test_larmor_201510_one_angle_smallish` folder which produced a singular file just within one minute, `test_stack_test_larmor_201510_one_angle` took around 1 and half minute whereas `test_stack_test_phantom1_3angles` took around 15 minutes for completion of the algorithm on each angle so total of 40 minutes. If a particular numbered file is not found within the folder, the algorithm will continue taking as long as it does to process and throw an error upon completion.. The algorithm works well even though it is still being developed. This PR is ready to be merged.

:shipit:

**Comment by FedeMPouzols** [comment]
@mantidproject/gatekeepers : this PR has been waiting for gatekeeping for 8 days now. There are other PRs that have been in "shipit" status also for one week or so.

**Comment by martyngigg** [comment]
Sorry, @FedeMPouzols. This does look fine.

**Pull request merged**